gnd/src/main/java/com/google/android/gnd/ui/editrecord/EditRecordViewModel.java

```diff
@@ -78,6 +86,47 @@
     this.showUnsavedChangesDialogEvents = new SingleLiveEvent<>();
     this.showErrorDialogEvents = new SingleLiveEvent<>();
     this.authManager = authenticationManager;
+    this.editRecordRequests = PublishSubject.create();
+    this.recordSaveRequests = PublishSubject.create();
+
+    disposeOnClear(
+            recordSaveRequests.zipWith(authManager.getUser(), this::saveRecord)
```

Comment:
Doesn't this try to save the record as soon as the viewmodel is instantiated? Doesn't sound like what we want to do here.

Reply:
(long reply ahead)

In this case it actually only calls `saveRecord` on a click, but I will admit that while this is clear on the surface and I think(?) the best transformation for the job as far as RX semantics are concerned, if I attempt to reason about what's actually going on here it's quite difficult. To make certain, I've gone ahead and double checked this manually by running this and logging (via `doOnNext`) on the zipped stream. The subscription here does indeed only fire on save button clicks.

There is a better approach here, but it's mostly due to the fact that we have to deal with streams that cross contextual boundaries, it's not an issue inherent to the use of `zipWith`.

One thing that's important to keep in mind about zips:

>The resulting {@code ObservableSource<R>} returned from {@code zip} will invoke {@link Observer#onNext onNext}
     * as many times as the number of {@code onNext} invocations of the source ObservableSource that emits the fewest
     * items.

So, completions of the respective observables *do matter*. Basically, whichever one of the sources completes first will also cause the resulting zipped stream to complete.

The first piece, `recordSaveRequests` is easy enough to comprehend. It's defined in the same file, and it's `onNext` is called on saves. This means that the zipped stream will propagate emissions from the first observable whenever `saveChanges()` is called. That makes sense. `zipWith` subscribes to each stream in order, so first, it will subscribe to `recordSaveRequests`, which means that this stream will effectively control the flow of the zipped stream. `recordSaveRequests` is only called in this file, and we're confident it does not call `onComplete` anywhere (just like the zipped stream, it just lives as long as the VM)  which means it won't ever shortcircuit emissions from the other observable.

It's more difficult to reason about the `authManager` part since the source observable is coming from elsewhere. It turns out that the observable being returned is actually a  `BehaviorSubject`, but several of the methods that return a reference to it are typed generically as `Observables`. Because it's a behavior subject, we can actually be confident that we'll always have a value to pass to our binop, since the behavior will replay the previous value upon initial subscription--other than this initial value however, I don't believe any other value comes from this source in this particular case. There's actually a major problem in our reference to this stream since we *have no idea when or if it will complete* from the context of this VM, so it could potentially kill our zipped stream by completing it at any time--we need to step outside and peek, potentially at the whole code base, to determine when or if this stream will ever kill or complete our zipper. Luckily, at the moment it is fine, but this is pretty bad situation since it's not guaranteed.  

The problems related to `authManager` actually hold, so far as I can tell, for *any* observable that's coming from a different scope/context/class. Thus far, I can think of two ways of handling potential completions or errors from external contexts, we do a bit of both of these things at present, more frequently relying on (1.)

1. Handle all errors and completions in the original context, and wrap results in a Resource or Result, giving other contexts a means of handling completions or errors through value checking. 
2. Use RX semantics to handle this, e.g. use `doOnComplete` etc. in situations in which the completion of a stream brought in from another context matters, e.g. when using `zipWith`.

The problem with approach (1.) is that this doesn't stop certain RX mechanisms from firing. For example, if we took  this approach for the situation described here, even though we're returning a wrapper object, e.g. `Resource<User>` which the `saveRecrod` function can us to get a sense of whether or not the origin stream completed, the original stream's `onComplete` *can still be called* resulting in the termination of the zipped stream regardless. Contrarily, `onErrorReturn` actually *does* prevent `onError` from being called, letting the stream continue indefinitely or continue to completion--but this is somewhat problematic as it alters the semantics of the observable significantly. That said, this approach eliminates the need for `zipWith`, but the unknown call to `onComplete` can still be problematic.

As for (2.) it's more verbose in most cases, and we don't (?) yet have good markers for things like "Observable that never completes." or "Observable that never errors." or "Observable that completes and errors normally" which would serve as huge hints in handling observers coming from other contexts appropriately. Another option is to simply handle each case always, ensuring that we use robust subscribers (defining onNext, onError, and onComplete handlers) and handling completions and errors appropriately in stream transformations as well (e.g. using `doOnDispose` when using `zipWith`. 

In general, this scenario suggests it's optimal to keep RX entities as self contained and bounded as possible. We're able to make very confident assertions about the `recordSaveRequest` subject because it's entirely defined and manipulated in a single scope/context. If we need to reference such a stream from outside this context, we have no idea whether errors are being handled, whether the stream completes, etc. It also suggests to me that it's best to define all three callbacks on subscribe(onNext, onError, onComplete).

Sooooo, we need `Result` for 'interior' streams that subvert `onError` to continue emitting items. But, to handle errors on 'exterior' streams we should probably just use subscribe callbacks?

This is still a problematic setup, I think, because as an implementer, there's no way for me to know whether a given stream calls `onErrorReturn` unless I look at the implementation, or unless I'm calling it myself on an interior stream using `Result.mapX` -- but I don't need to call this function if the original stream is already calling `onErrorReturn` which I can't know unless, again, I check the implementation.

For many of the functions that return observables that are fetching some data from the db, we can probably ditch `onErrorReturn` in favor of forms of `retry` and error handling in `subscribe()` which would remove the need for some wrappers and disambiguate some streams (i.e. does this stream potentially error or not).

lets chat about this since my hesd's muddle enough at this point haha